drivers/net/ethernet/intel/idpf/idpf_eth_idc.c

```diff
@@ -0,0 +1,453 @@
+								 vport_id);
+			if (eth_info)
+				eth_idc_drv->event_handler(eth_info, &event);
+			break;
+
+		default:
+			break;
+		}
+		break;
+
+	default:
+		break;
+	}
+}
+
+/**
+ * idpf_eth_idc_init_shared - Initialize ethernet shared struct
+ * @eth_shared: Ethernet and main apdater shared struct
+ * @pdev: Primary PCI device information struct
+ *
+ * Returns 0 on success, negative on failure
+ */
+int idpf_eth_idc_init_shared(struct idpf_eth_shared *eth_shared,
+			     struct pci_dev *pdev)
+{
+	eth_shared->pdev = pdev;
+	/* IDC ops */
+	eth_shared->eth_idc_ops.event_send = idpf_eth_idc_event_send;
+	eth_shared->eth_idc_ops.virtchnl_send = idpf_eth_idc_virtchnl_send;
+	eth_shared->eth_idc_ops.intr_reg_init = idpf_eth_idc_intr_reg_init;
+	eth_shared->eth_idc_ops.intr_init_vec_idx =
+		idpf_eth_idc_intr_init_vec_idx;
+	eth_shared->eth_idc_ops.req_rel_vec_idx =
+		idpf_eth_idc_req_rel_vector_indexes;
+
+	return 0;
+}
+
+/**
+ * idpf_eth_idc_deinit_shared - Initialize ethernet shared struct
+ * @eth_shared: Ethernet and main apdater shared struct
+ */
+void idpf_eth_idc_deinit_shared(struct idpf_eth_shared *eth_shared)
+{
+	eth_shared->pdev = NULL;
+	eth_shared->hw_addr = NULL;
+}
+
+/**
+ * idpf_eth_idc_device_free - function to be mapped to aux dev's release op
+ * @dev: pointer to device to free allocated memory
+ */
+static
+void idpf_eth_idc_device_free(struct device *dev)
+{
+	struct idpf_eth_idc_auxiliary_dev *eth_adev;
+	struct idpf_adapter *adapter;
+	u16 idx;
+
+	eth_adev = container_of(dev, struct idpf_eth_idc_auxiliary_dev,
+				adev.dev);
+	adapter = pci_get_drvdata(to_pci_dev(eth_adev->adev.dev.parent));
+	ida_free(&idpf_eth_idc_ida, eth_adev->adev.id);
+	idx = eth_adev->eth_info.idx;
+	/* Release all max queues allocated to the pool */
+	idpf_dealloc_max_qs(adapter, &eth_adev->eth_info.caps.q_info);
+	kfree(adapter->adevs[idx]);
+	adapter->adevs[idx] = NULL;
+}
+
+/**
+ * idpf_eth_idc_driver_register - Register ethernet auxiliary driver
+ * @void: void
+ *
+ * Returns 0 on success, negative on failure
+ */
+int idpf_eth_idc_driver_register(void)
+{
+	struct idpf_eth_idc_auxiliary_driver *eth_idc_drv;
+
+	eth_idc_drv = idpf_eth_idc_get_driver();
+	return auxiliary_driver_register(&eth_idc_drv->adrv);
+}
+
+/**
+ * idpf_eth_idc_driver_unregister - unregister ethernet driver
+ * @void: void
+ */
+void idpf_eth_idc_driver_unregister(void)
+{
+	struct idpf_eth_idc_auxiliary_driver *eth_idc_drv;
+
+	eth_idc_drv = idpf_eth_idc_get_driver();
+	auxiliary_driver_unregister(&eth_idc_drv->adrv);
+}
+
+/**
+ * idpf_eth_idc_init_device_params - Initialize device params
+ * @adapter: Idpf private structure
+ * @eth_dev: Ethernet dev info
+ */
+static void
+idpf_eth_idc_init_device_params(struct idpf_adapter *adapter,
+				struct idpf_eth_idc_auxiliary_dev *eth_dev)
+{
+	/* crc info */
+	eth_dev->eth_info.caps.crc_enable = adapter->crc_enable;
+
+	/* Queue Info */
+	eth_dev->eth_info.caps.q_info.max_rxq =
```

Comment:
This is taken care as part of alloc_max_qs in the calling function right??

Reply:
Initially PCI driver only passing those max values to aux device. IDC then pre-allocate for the aux device.